Jaeger Zipkin OpenTelemetry guides (#994)
micronaut-projects/micronaut-guides: guides/micronaut-microservices-distributed-tracing-jaeger-opentelemetry/micronaut-microservices-distributed-tracing-jaeger-opentelemetry.adoc

common:header-top.adoc[]

== Getting Started

> In this guide, we will integrate https://www.jaegertracing.io[Jaeger] in a Micronaut application composed of three microservices.

Reviewer: ```suggestion
In this guide, we will integrate https://www.jaegertracing.io[Jaeger] with a Micronaut application composed of three microservices.
```

image::flow.svg[]

A request to `bookrecommendation` (http://localhost:8080/books) triggers several requests through our microservices mesh.

common:annotationprocessors.adoc[]

== Jaeger and the Micronaut Framework

=== Install Jaeger via Docker

The quickest way to start https://www.jaegertracing.io[Jaeger] is via Docker:

[source, bash]
----
docker run -d --name jaeger \
  -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 \
  -p 5775:5775/udp \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 14268:14268 \
  -p 14250:14250 \
  -p 9411:9411 \
  jaegertracing/all-in-one:latest
----

== OpenTelemetry dependencies

common:opentelemetry-annotations.adoc[]

common:opentelemetry-http.adoc[]

common:opentelemetry-exporter.adoc[]

Each service adds the OpenTelemetry Jaeger exporter dependency.

dependency:opentelemetry-exporter-jaeger[groupId=io.opentelemetry,scope=compile]

Each service sets `jaeger` as exporter in configuration:

resource:application.yml[app=bookcatalogue,tag=jaeger]

common:opentelemetry-disable.adoc[]

=== Book Inventory Changes

Annotate the method with `@ContinueSpan` and the parameter with `@SpanTag`:

source:BooksController[app=bookinventory]

> callout:continue-span[1]

Reviewer 2: Since the `@ContinueSpan` is ignored by the MN OT code, the callout
> The @ContinueSpan annotation will continue an existing span, wrapping the method call or reactive type.

Is an incorrect statement.

> callout:span-tag[2]

Reviewer 2: With MN OT code, the second sentence of the callout
> When you use @SpanTag you need either to annotate the method with @NewSpan or @ContinueSpan.

Is an incorrect statement.

Reviewer 2: I was wrong, ignore comment.

:exclude-for-build:

[source,bash]
----
...
14:28:34.034 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 499ms. Server Running: http://localhost:8081
----

Run `bookinventory` microservice:

:exclude-for-build:maven

To run the application, execute `./gradlew run`.

:exclude-for-build:

:exclude-for-build:gradle

To run the application, execute `./mvnw mn:run`.

:exclude-for-build:

[source,bash]
----
...
14:31:13.104 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 506ms. Server Running: http://localhost:8082
----

Run `bookrecommendation` microservice:

:exclude-for-build:maven

To run the application, execute `./gradlew run`.

:exclude-for-build:

:exclude-for-build:gradle

To run the application, execute `./mvnw mn:run`.

:exclude-for-build:

[source,bash]
----
...
14:31:57.389 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 523ms. Server Running: http://localhost:8080
----

You can run a cURL command to test the whole application:

[source, bash]
----
curl http://localhost:8080/books
----

[source,json]
----
[{"name":"Building Microservices"}
----

You can then navigate to http://localhost:16686 to access the Jaeger UI.

> The previous request generates a traces composed by 9 spans.

Reviewer: ```suggestion
The previous request generates the following traces, composed of 9 spans.
```

image::jaegerui-opentelemetry.png[]

In the previous image, you can see that:

- Whenever a Micronaut HTTP client executes a new network request, it creates a new span.
- Whenever a Micronaut server receives a request, it creates a new span.

The `stock.isbn` tags that we configured with `@SpanTag` are present.

Moreover, you can see the requests to `bookinventory` are done in parallel.

== Next Steps

> As you have seen in this guide, without any annotations, you get distributed tracing up and running fast with the Micronaut framework.

Reviewer: ```suggestion
As you have seen in this guide, without any annotations, you can get distributed tracing up and running quickly with the Micronaut framework.
```
